beacon-chain/sync/data_columns_sampling.go

```diff
@@ -4,231 +4,254 @@ import (
-	// Retrieve the custody count of the peer.
-	custodySubnetCount := s.cfg.p2p.CustodyCountFromRemotePeer(pid)
+	// custodyColumns is a set of columns that the node is responsible for custody.
+	custodyColumns map[uint64]bool
+	// columnFromPeer maps a peer to the columns it is responsible for custody.
+	columnFromPeer map[peer.ID]map[uint64]bool
+	// peerFromColumn maps a column to the peer responsible for custody.
+	peerFromColumn map[uint64]map[peer.ID]bool
+}
 
-	// Extract the node ID from the peer ID.
-	nodeID, err := p2p.ConvertPeerIDToNodeID(pid)
-	if err != nil {
-		return nil, errors.Wrap(err, "extract node ID")
+// newDataColumnSampler1D creates a new 1D data column sampler.
+func newDataColumnSampler1D(
+	p2p p2p.P2P,
+	clock *startup.Clock,
+	ctxMap ContextByteVersions,
+	stateNotifier statefeed.Notifier,
+) *dataColumnSampler1D {
+	columnToPeerMap := make(map[uint64]map[peer.ID]bool, params.BeaconConfig().NumberOfColumns)
+	for i := uint64(0); i < params.BeaconConfig().NumberOfColumns; i++ {
+		columnToPeerMap[i] = make(map[peer.ID]bool)
 	}
 
-	// Determine which columns the peer should custody.
-	custodyColumns, err := peerdas.CustodyColumns(nodeID, custodySubnetCount)
-	if err != nil {
-		return nil, errors.Wrap(err, "custody columns")
+	return &dataColumnSampler1D{
+		p2p:            p2p,
+		clock:          clock,
+		ctxMap:         ctxMap,
+		stateNotifier:  stateNotifier,
+		columnFromPeer: make(map[peer.ID]map[uint64]bool),
+		peerFromColumn: columnToPeerMap,
 	}
-
-	return custodyColumns, nil
 }
 
-// sampleDataColumnsFromPeer samples data columns from a peer.
-// It filters out columns that were not requested and columns with incorrect root.
-// It returns the retrieved columns.
-func (s *Service) sampleDataColumnsFromPeer(
-	pid peer.ID,
-	requestedColumns map[uint64]bool,
-	root [fieldparams.RootLength]byte,
-) (map[uint64]bool, error) {
-	// Build the data column identifiers.
-	dataColumnIdentifiers := make(types.DataColumnSidecarsByRootReq, 0, len(requestedColumns))
-	for index := range requestedColumns {
-		dataColumnIdentifiers = append(dataColumnIdentifiers, &eth.DataColumnIdentifier{
-			BlockRoot:   root[:],
-			ColumnIndex: index,
-		})
+// Run implements DataColumnSampler.
+func (d *dataColumnSampler1D) Run(ctx context.Context) {
+	// verify if we need to run sampling or not, if not, return directly
+	csc := peerdas.CustodySubnetCount()
+	columns, err := peerdas.CustodyColumns(d.p2p.NodeID(), csc)
+	if err != nil {
+		log.WithError(err).Error("Failed to determine local custody columns")
+		return
 	}
+	d.custodyColumns = columns
 
-	// Send the request.
-	roDataColumns, err := SendDataColumnSidecarByRoot(s.ctx, s.cfg.clock, s.cfg.p2p, pid, s.ctxMap, &dataColumnIdentifiers)
-	if err != nil {
-		return nil, errors.Wrap(err, "send data column sidecar by root")
+	custodyColumnsCount := uint64(len(columns))
+	if custodyColumnsCount > params.BeaconConfig().NumberOfColumns/2 {
```

Comment:
With `n` the number of columns:

- If `n` is even, then `n/2` is enough to reconstruct the whole matrix. ==> `>` can be replaced by `>=`.
- If `n` is odd, then we `n/2+1` is enough to reconstruct the whole matrix. ==> We need to keep `>`.

Even if it is very unlikely to have a odd number of columns, the code should handle all the cases.